src/backend/distributed/connection/shared_connection_stats.c

```diff
@@ -380,177 +476,167 @@ TryToIncrementSharedConnectionCounter(const char *hostname, int port)
 		 * a reasonable pace. The latter limit typically kicks in when the database
 		 * is issued lots of concurrent sessions at the same time, such as benchmarks.
 		 */
-		if (activeBackendCount + 1 > GetLocalSharedPoolSize())
+		bool localNodeLimitExceeded =
+			connectionToLocalNode &&
+			(GetLocalSharedPoolSize() == DISABLE_REMOTE_CONNECTIONS_FOR_LOCAL_QUERIES ||
+			 GetExternalClientBackendCount() + 1 > currentConnectionsLimit);
+		if (remoteNodeLimitExceeded || localNodeLimitExceeded)
 		{
-			counterIncremented = false;
+			connectionSlotAvailable = false;
 		}
-		else if (connectionEntry->connectionCount + 1 > GetLocalSharedPoolSize())
+	}
+
+	if (connectionSlotAvailable)
+	{
+		if (maintenanceConnection)
 		{
-			counterIncremented = false;
+			workerNodeConnectionEntry->maintenanceConnectionsCount += 1;
 		}
 		else
 		{
-			connectionEntry->connectionCount++;
-			counterIncremented = true;
+			workerNodeConnectionEntry->regularConnectionsCount += 1;
 		}
+		workerNodeDatabaseEntry->count += 1;
 	}
-	else if (connectionEntry->connectionCount + 1 > GetMaxSharedPoolSize())
-	{
-		/* there is no space left for this connection */
-		counterIncremented = false;
-	}
-	else
+
+	if (IsLoggableLevel(DEBUG4))
```

Comment:
I thought it was there to prevent building of the expensive `errmsg`. Maybe it is better to leave if this way just in case, to avoid some performance regression?